### PR TITLE
Optimized the backward convolution function

### PR DESCRIPTION
**Pull request by MarkFischinger**
I've made several optimizations to the backward function in the convolution code; the main areas of improvement include padding, rotation, and dilation operations. These changes aim to reduce computational overhead and a speed up.

**Changes**

_Padding Optimization_
Instead of adding padding to the input data separately, I've integrated the fixed padding values directly into the convolution operation. This reduces the number of steps required.

_Rotation Optimization_
The previous method involved rotating filters by 180 degrees. To optimize this, I replaced the rotation with a combination of matrix flipping or transposing followed by reversing. Additionally I embedded the rotation directly into the convolution step, which simplifies the overall computation.

_Dilation Optimization_
For consistent and predetermined dilation factors, I've embedded these directly into the convolution. This minimizes the total computational workload and enhances performance.

I'm happy to show you these benchmarks run based on the mnist-code from the [mlpack-benchmarks repo](https://github.com/MarkFischinger/mlpack-benchmarks/blob/master/src/convolutional/backward/mnist/mnist_benchmark.cpp):
```
Time Elapsed: 73222ms (original)
Time Elapsed: 50517ms (optimized)
```

### CODE REVIEW

File: src/mlpack/methods/ann/layer/convolution_impl.hpp [hunk 1]
```diff
@@ -382,41 +382,34 @@ void ConvolutionType<
       inMaps * higherInDimensions * batchSize);
   gTemp.zeros();
 
-  const bool usingPadding =
-      (padWLeft != 0 || padWRight != 0 || padHTop != 0 || padHBottom != 0);
-
-  // To perform the backward pass, we need to rotate all the filters.
-  CubeType rotatedFilters(weight.n_rows,
-      weight.n_cols, weight.n_slices);
+  const bool usingPadding = (padWLeft != 0 || padWRight != 0 || padHTop != 0 || padHBottom != 0);
```
Comment (shrit) [inline src/mlpack/methods/ann/layer/convolution_impl.hpp:385]: please keep the line under 80 characters, basically identical to how it was

File: src/mlpack/methods/ann/layer/convolution_impl.hpp [hunk 2]
```diff
@@ -382,41 +382,34 @@ void ConvolutionType<
...
-  CubeType dilatedMappedError;
-  if (strideHeight == 1 && strideWidth == 1)
-  {
-    MakeAlias(dilatedMappedError, mappedError, mappedError.n_rows,
-        mappedError.n_cols, mappedError.n_slices);
-  }
-  else
+  CubeType dilatedMappedError(mappedError.n_rows * (strideWidth == 1 ? 1 : strideWidth - 1),
+                              mappedError.n_cols * (strideHeight == 1 ? 1 : strideHeight - 1),
+                              mappedError.n_slices);
```
Comment (shrit) [inline src/mlpack/methods/ann/layer/convolution_impl.hpp:389]: The same for this one, please keep it under 80 chars

**Review by shrit — CHANGES_REQUESTED** [review]
The `GradientConvolutionLayerWithStrideTest` is failing, is this happening as well on your machine? I did not run this locally, could you look into this one ?

**Comment by MarkFischinger** [comment]
@shrit After a lot testing/debugging, I believe I have finally found the issue. This should resolve it.

File: src/mlpack/methods/ann/layer/convolution_impl.hpp [hunk 3]
```diff
@@ -427,31 +424,25 @@ void ConvolutionType<
...
+              BackwardConvolutionRule::Convolution(
+                  dilatedMappedError.slice(outMap + offset * maps),
+                  rotatedFilters.slice((outMap * inMaps) + inMap),
+                  curG,
+                  strideWidth, 
+                  strideHeight, 
+                  1, 
+                  1, 
+                  true); 
+          }
```
[the remark this answers is not in the document]
Comment (MarkFischinger) [reply src/mlpack/methods/ann/layer/convolution_impl.hpp:444]: Updated my IDE configuration. This shouldn't happen again :)

**Comment by MarkFischinger** [comment]
So, I found out that these small adjustments increased the speed by approximately 30.5%.

Here are the MNIST benchmarks:
**Original**
```
756/756 [====================================================================================================] 100% - 59.914s/epoch; 79ms/step; loss: 9.42397
Validation loss: 10191.9.
Time elapsed: 61458ms
Accuracy: train = 70.4048%,      valid = 70.1429%
```


**Optimized**
```
756/756 [====================================================================================================] 100% - 41.1419s/epoch; 54ms/step; loss: 9.42397
Validation loss: 10191.9.
Time elapsed: 42741ms
Accuracy: train = 70.4048%,      valid = 70.1429%
```

**Review by rcurtin — APPROVED** [review]
Looks good!  The speedup is nice.

**Review by zoq — APPROVED** [review]

**Review by shrit — APPROVED** [review]

**Pull request merged**